components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/impl/CertificateBasedTokenBinder.java

```diff
@@ -0,0 +1,196 @@
+
+        return Optional.ofNullable(generateCnfHashValue(oAuth2AccessTokenReqDTO.getHttpServletRequestWrapper()));
+    }
+
+    @Override
+    public void setTokenBindingValueForResponse(HttpServletResponse response, String bindingValue) {
+
+        // Not required.
+    }
+
+    @Override
+    public void clearTokenBindingElements(HttpServletRequest request, HttpServletResponse response) {
+
+        // Not required.
+    }
+
+    @Override
+    public boolean isValidTokenBinding(Object request, String bindingReference) {
+
+        String cnfValue = generateCnfHashValue((HttpServletRequest) request);
+        if (StringUtils.isNotBlank(cnfValue)) {
+            return StringUtils.equals(bindingReference, OAuth2Util.getTokenBindingReference(cnfValue));
+        }
+        return false;
+    }
+
+    @Override
+    public boolean isValidTokenBinding(OAuth2AccessTokenReqDTO oAuth2AccessTokenReqDTO, String bindingReference) {
+
+        String refreshToken = oAuth2AccessTokenReqDTO.getRefreshToken();
+        try {
+            TokenBinding tokenBinding = OAuthTokenPersistenceFactory.getInstance().getTokenBindingMgtDAO()
+                    .getBindingFromRefreshToken(refreshToken, OAuth2Util.isHashEnabled());
+            String cnfValue = generateCnfHashValue(oAuth2AccessTokenReqDTO.getHttpServletRequestWrapper());
+
+            if (tokenBinding != null && CERTIFICATE_BASED_TOKEN_BINDER.equals(tokenBinding.getBindingType()) &&
+                    StringUtils.isNotBlank(cnfValue)) {
+                return StringUtils.equals(cnfValue, tokenBinding.getBindingValue()) &&
+                        StringUtils.equals(bindingReference, tokenBinding.getBindingReference());
+            }
+            return false;
+        } catch (IdentityOAuth2Exception e) {
+            return false;
+        }
+    }
+
+    private String generateCnfHashValue(HttpServletRequest request) {
+
+        Base64URL certThumbprint;
+        X509Certificate certificate = null;
+        String headerName = Optional.ofNullable(IdentityUtil.getProperty(OAuthConstants.MTLS_AUTH_HEADER))
+                .orElse(OAuthConstants.CONFIG_NOT_FOUND);
+
+        String certificateInHeader = request.getHeader(headerName);
+        Object certObject = Optional.ofNullable(request.getAttribute(OAuthConstants.JAVAX_SERVLET_REQUEST_CERTIFICATE))
+                .orElse(null);
+
+        if (StringUtils.isNotBlank(certificateInHeader)) {
+            try {
+                certificate = parseCertificate(certificateInHeader);
+            } catch (CertificateException e) {
+                /* Adding a debug log as these errors cannot be thrown as per the TokenBinder interface implementation.
+                   But null checks have been performed where these methods are being executed. */
+                log.debug("Error occurred while extracting the certificate from the request header.", e);
```

Comment:
Lets wrap this in a log.isDebugEnabled()

Reply:
Fixed in 3da41018b6a5b63ad5435d77db06c498340e3af0